onnxruntime/core/external_ops/pyop.cc

```diff
@@ -0,0 +1,219 @@
+    if (PyErr_Occurred()) {
+        PyObject *type, *value, *trace;
+        PyErr_Fetch(&type, &value, &trace);
+        if (nullptr != value) {
+            auto pyVal = PyObject_Repr(value);
+            scope.Add(pyVal);
+            auto pyStr = PyUnicode_AsEncodedString(pyVal, "utf-8", "Error ~");
+            scope.Add(pyStr);
+            err = PyBytes_AS_STRING(pyStr);
+        }
+        PyErr_Restore(type, value, trace);
+    }
+    return err.c_str();
+}
+
+PyObject* MakePyObj (const void* data, int32_t type, const vector<int64_t>& dim)
+{
+    std::vector<npy_intp> np_dim;
+    for (auto d: dim) {
+        np_dim.push_back(static_cast<npy_intp>(d));
+    }
+    auto pyObj = static_cast<PyObject*>(PyArray_EMPTY(static_cast<int>(np_dim.size()), np_dim.data(), type, 0));
+    auto data_len = std::accumulate(begin(np_dim), end(np_dim),
+                                    static_cast<int64_t>(PyArray_DescrFromType(type)->elsize),
+                                    std::multiplies<int64_t>());
+    auto np_array = reinterpret_cast<PyArrayObject*>(pyObj);
+    memcpy(PyArray_DATA(np_array), data, data_len);
+    return pyObj;
+}
+
+bool ExtractOutput (PyObject*                pyObj,
+                    vector<const void*>&     outputs,
+                    vector<int32_t>&         outputs_elem_size,
+                    vector<vector<int64_t>>& outputs_dim)
+{
+    if (!PyArray_Check(pyObj)) {
+        return false;
+    }
+
+    outputs_dim.push_back({});
+    auto np_array = reinterpret_cast<PyArrayObject*>(pyObj);
+    outputs_elem_size.push_back(static_cast<int32_t>(PyArray_ITEMSIZE(np_array)));
+
+    for (int i = 0; i < PyArray_NDIM(np_array); ++i) {
+        outputs_dim.back().push_back(PyArray_SHAPE(np_array)[i]);
+    }
+
+    auto data_len = std::accumulate(begin(outputs_dim.back()),
+                                    end(outputs_dim.back()),
+                                    static_cast<int64_t>(outputs_elem_size.back()),
+                                    std::multiplies<int64_t>());
+
+    outputs.push_back(new char[data_len]);
+    memcpy(const_cast<void*>(outputs.back()), PyArray_DATA(np_array), data_len);
+    return true;
+}
+
+PYOP_EXPORT void* NewInstance (const char* module, const char* class_name, const unordered_map<string, string>& args)
+{
+    Scope scope; 
+    auto pyModule = PyImport_ImportModule(module);
+    if (nullptr == pyModule) {
+        return nullptr;
+    }
+
+    scope.Add(pyModule);
+    auto pyClass  = PyObject_GetAttrString(pyModule, class_name);
+    if (nullptr == pyClass) {
+        return nullptr;
+    }
+
+    scope.Add(pyClass);
+    auto empty_args = PyTuple_New(0);
+    scope.Add(empty_args);
+    auto named_args = PyDict_New();
+    scope.Add(named_args);
+    for (const auto& iter: args) {
+        PyDict_SetItemString(named_args, iter.first.c_str(), PyUnicode_FromString(iter.second.c_str()));
+    }
+
+    return PyObject_Call(pyClass, empty_args, named_args);
```

Comment:
You chose to use Scope class to call DECREF on every object before leaving the function. What happens to object created in this function, used somewhere else? Usually, developpers add DECREF in the function because the code is easier to read. I think one DECREF is missing of the return of PyObject_Call to be added in InvokePythonFunc.

Reply:
They get released on spot because they are expected no longer.
For the one not released, it is to be consumed later - it is a instance of a Python class whose "compute" function will be triggered later.